core/roboconf-target-embedded/src/main/java/net/roboconf/target/embedded/internal/EmbeddedHandler.java

```diff
@@ -84,7 +115,87 @@ public boolean isMachineRunning( TargetHandlerParameters parameters, String mach
 	@Override
 	public String retrievePublicIpAddress( TargetHandlerParameters parameters, String machineId )
 	throws TargetException {
-		// This handler cannot determine the IP address
+		String ip = this.machineIdToRunning.get(machineId);
+		return Utils.isEmptyOrWhitespaces(ip) ? null : ip;
+	}
+
+	/**
+	 * Acquires an IP address among available ones.
+	 * @return The acquired IP address
+	 */
+	protected String acquireIpAddress() {
+		if(ipTable != null) {
+			for(Map.Entry<String, Boolean> entry : ipTable.entrySet()) {
+				if(! entry.getValue()) {
+					entry.setValue(true);
+					return entry.getKey();
+				}
+			}
+		}
 		return null;
 	}
+
+	/**
+	 * Release a (previously acquired) IP address.
+	 * @param ip The IP address to release
+	 */
+	protected void releaseIpAddress(String ip) {
+		if(ipTable != null && ipTable.get(ip) != null) {
+			ipTable.put(ip, false);
+		}
+	}
+
+	/**
+	 * Send configuration to remote host using scp. A user-data file will be generated,
+	 * sent to the remote host (in agent configuration directory), then the remote agent configuration file
+	 * will be updated to point on the new user-data file ("parameters" property).
+	 * @param ip IP address of remote host
+	 * @param parameters Target handler parameters (with user-data inside)
+	 * @throws IOException
+	 */
+	protected void sendConfiguration(String ip, TargetHandlerParameters parameters) throws IOException {
+		SSHClient ssh = new SSHClient();
+		try {
+			ssh.loadKnownHosts();
+			ssh.connect(ip);
+
+			String user = parameters.getTargetProperties().get(SCP_USER);
+			if(user == null) user = "ubuntu"; // default on several (ubuntu) systems, including IaaS VMs
+			String keyfile = parameters.getTargetProperties().get(SCP_KEYFILE);
+
+			if(keyfile == null) ssh.authPublickey(user); // use ~/.ssh/id_rsa and ~/.ssh/id_dsa
+			else ssh.authPublickey(user, keyfile); // load key from specified file (eg .pem).
+
+			String userData = UserDataHelpers.writeUserDataAsString(
+					parameters.getMessagingProperties(),
+					parameters.getDomain(),
+					parameters.getApplicationName(),
+					parameters.getScopedInstancePath());
+
+			// Generate local user-data file, then upload it
+			File userdataFile = new File("/tmp", USER_DATA_FILE);
+			Utils.writeStringInto(userData, userdataFile);
+			ssh.newSCPFileTransfer().upload(new FileSystemFile(userdataFile), AGENT_CONFIG_DIR);
+
+			File localAgentConfig = new File("/tmp", "net.roboconf.agent.configuration.cfg");
```

Comment:
I am not a fan of `/tmp` in the code.
Use `System.getProperty( "java.io.tmpdir" )`.